### PR TITLE
Extract SharedBlobCacheService LFU logic behind a private cache interface

### PR DESCRIPTION
**Pull request by romseygeek**
This commit is a first step to adding an alternative LRU cache implementation, while
keeping the initial implementation available for comparison testing.

### CODE REVIEW

**Comment by romseygeek** [comment]
@elasticmachine update branch

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 1]
```diff
@@ -408,34 +399,66 @@ private int getRegionSize(long fileLength, int region) {
...
+            assignToSlot(entry, freeSlot);
+        } else {
+            // need to evict something
+            cache.evict();
+            final SharedBytes.IO freeSlotRetry = freeRegions.poll();
+            if (freeSlotRetry != null) {
+                assignToSlot(entry, freeSlotRetry);
+            } else {
+                synchronized (this) {
+                    cache.evictMatching(r -> r == regionKey);
```
Comment (original-brownbear) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:440]: This is a potentially massive slowdown relative to the previous version doing a map remove instead isn't it? We'd have to iterate all the pages to do this? (which would be uncool if we wanted to have more of those and smaller ones in the future :))

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 2]
```diff
@@ -269,13 +268,10 @@ public void validate(ByteSizeValue value, Map<Setting<?>, Object> settings, bool
 
     private final int numRegions;
     private final ConcurrentLinkedQueue<SharedBytes.IO> freeRegions = new ConcurrentLinkedQueue<>();
```
Comment (original-brownbear) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:270]: Should we push the free-list into the cache interface as well maybe? Then we could do away with the `evict` call and instead do `evictAndGet` on the cache directly?

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java [hunk 3]
```diff
@@ -1189,4 +979,259 @@ public record Stats(
...
+    private interface Cache<K, V> {
+        V get(K key);
+
+        V computeIfAbsent(K key, Function<K, V> computeFunction);
+
+        void evict();
+
+        int evictMatching(Predicate<K> predicate);
+
+        void close();
```
Comment (original-brownbear) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java:992]: We could also just extend `Releasable` here on the interface I think.

**Review by original-brownbear — COMMENTED** [review]
This looks like just what we need. Just one suggestion to go a little further here and one problem with performance I think, but other than that looks really nice! Thanks

**Review by kc13greiner — APPROVED** [review]
CODEOWNERS LGTM  since`.fleet-` are system indexes

**Review by kpollich — APPROVED** [review]
Fleet changes

**Comment by romseygeek** [comment]
Apologies to fleet people, I have no idea what happened there...

**Comment by romseygeek** [comment]
Closing in favour of #100528

**Pull request closed without merging**